File: web-sdks/dart/lib/src/auth/auth_models.dart
/// Authentication models for GroupVAN SDK
///
/// Models the JWT-based authentication system used by GroupVAN API.
/// Supports access tokens, refresh tokens, and automatic token management.
library auth_models;

import 'package:json_annotation/json_annotation.dart';
import 'package:equatable/equatable.dart';
import 'package:meta/meta.dart';

import '../models/auth.dart' show User;
part 'auth_models.g.dart';

/// Login request model
@JsonSerializable()
class LoginRequest extends Equatable {
  /// Email for authentication
  final String email;

  /// Password for authentication
  final String password;

  const LoginRequest({required this.email, required this.password});

  factory LoginRequest.fromJson(Map<String, dynamic> json) =>
      _$LoginRequestFromJson(json);

  Map<String, dynamic> toJson() => _$LoginRequestToJson(this);

  @override
  List<Object?> get props => [email, password];
}

/// Token response from authentication endpoints
///
/// Note: refresh_token is no longer returned in API responses.
/// It is now set as an HttpOnly cookie by the server.
@JsonSerializable()
class TokenResponse extends Equatable {
  /// JWT access token for API requests
  @JsonKey(name: 'access_token')
  final String accessToken;

  /// Token expiration time in seconds
  @JsonKey(name: 'expires_in')
  final int expiresIn;

  /// Token type (typically 'Bearer')
  @JsonKey(name: 'token_type', defaultValue: 'Bearer')
  final String tokenType;

  const TokenResponse({
    required this.accessToken,
    required this.expiresIn,
    this.tokenType = 'Bearer',
  });

  factory TokenResponse.fromJson(Map<String, dynamic> json) =>
      _$TokenResponseFromJson(json);

  Map<String, dynamic> toJson() => _$TokenResponseToJson(this);

  @override
  List<Object?> get props => [accessToken, expiresIn, tokenType];
}

/// JWT token claims decoded from access token
@JsonSerializable()
class TokenClaims extends Equatable {
  /// User ID from the token
  @JsonKey(name: 'sub')
  final String userId;

  /// Token type (access/refresh)
  @JsonKey(name: 'type', defaultValue: 'access')
  final String type;

  /// Token issued at timestamp
  @JsonKey(name: 'iat')
  final int issuedAt;

  /// Token expiration timestamp
  @JsonKey(name: 'exp')
  final int expiration;

  /// Token JTI (unique identifier)
  @JsonKey(name: 'jti')
  final String jti;

  /// Member ID if present
  @JsonKey(name: 'member')
  final String? member;

  const TokenClaims({
    required this.userId,
    required this.issuedAt,
    required this.expiration,
    required this.jti,
    this.type = 'access',
    this.member,
  });

  factory TokenClaims.fromJson(Map<String, dynamic> json) =>
      _$TokenClaimsFromJson(json);

  Map<String, dynamic> toJson() => _$TokenClaimsToJson(this);

  /// Whether the token is expired
  bool get isExpired =>
      DateTime.now().millisecondsSinceEpoch > expiration * 1000;

  /// Whether the token will expire within the given duration
  bool willExpireWithin(Duration duration) {
    final expirationTime = DateTime.fromMillisecondsSinceEpoch(
      expiration * 1000,
    );
    final threshold = DateTime.now().add(duration);
    return expirationTime.isBefore(threshold);
  }

  /// Time until token expires
  Duration get timeUntilExpiration {
    final expirationTime = DateTime.fromMillisecondsSinceEpoch(
      expiration * 1000,
    );
    final now = DateTime.now();
    if (expirationTime.isBefore(now)) {
      return Duration.zero;
    }
    return expirationTime.difference(now);
  }

  @override
  List<Object?> get props => [userId, type, issuedAt, expiration, jti, member];
}

/// Authentication state
enum AuthState {
  /// Not authenticated
  unauthenticated,

  /// Currently authenticating
  authenticating,

  /// Authenticated with valid tokens
  authenticated,

  /// Token refresh in progress
  refreshing,

  /// Authentication expired
  expired,

  /// Authentication failed
  failed,
}

/// Current authentication status
///
/// Note: refreshToken is no longer stored client-side on web.
/// The browser manages the refresh token via HttpOnly cookies.
@immutable
class AuthStatus extends Equatable {
  /// Current authentication state
  final AuthState state;

  /// Access token if authenticated
  final String? accessToken;

  /// Decoded token claims if available
  final TokenClaims? claims;

  final User? userInfo;

  /// Last authentication error if any
  final String? error;

  /// Additional metadata for the current state (e.g., error context)
  final Map<String, dynamic>? metadata;

  /// Timestamp of last successful authentication
  final DateTime? authenticatedAt;

  /// Timestamp of last token refresh
  final DateTime? refreshedAt;

  const AuthStatus({
    required this.state,
    this.accessToken,
    this.claims,
    this.userInfo,
    this.error,
    this.metadata,
    this.authenticatedAt,
    this.refreshedAt,
  });

  /// Create unauthenticated status
  const AuthStatus.unauthenticated()
    : state = AuthState.unauthenticated,
      accessToken = null,
      claims = null,
      userInfo = null,
      error = null,
      metadata = null,
      authenticatedAt = null,
      refreshedAt = null;

  /// Create authenticating status
  const AuthStatus.authenticating()
    : state = AuthState.authenticating,
      accessToken = null,
      claims = null,
      userInfo = null,
      error = null,
      metadata = null,
      authenticatedAt = null,
      refreshedAt = null;

  /// Create authenticated status
  AuthStatus.authenticated({
    required String accessToken,
    required TokenClaims claims,
    User? userInfo,
  }) : state = AuthState.authenticated,
       accessToken = accessToken,
       claims = claims,
       userInfo = userInfo,
       error = null,
       metadata = null,
       authenticatedAt = DateTime.now(),
       refreshedAt = null;

  /// Create refreshing status
  AuthStatus.refreshing({
    required String accessToken,
    required TokenClaims claims,
    required DateTime authenticatedAt,
    User? userInfo,
  }) : state = AuthState.refreshing,
       accessToken = accessToken,
       claims = claims,
       userInfo = userInfo,
       error = null,
       metadata = null,
       authenticatedAt = authenticatedAt,
       refreshedAt = null;

  /// Create expired status
  AuthStatus.expired({
    String? error,
    required String accessToken,
    required DateTime authenticatedAt,
    DateTime? refreshedAt,
    User? userInfo,
  }) : state = AuthState.expired,
       accessToken = accessToken,
       claims = null,
       userInfo = userInfo,
       error = error,
       metadata = null,
       authenticatedAt = authenticatedAt,
       refreshedAt = refreshedAt;

  /// Create failed status
  AuthStatus.failed({required String error, Map<String, dynamic>? metadata})
    : state = AuthState.failed,
      accessToken = null,
      claims = null,
      userInfo = null,
      error = error,
      metadata = metadata,
      authenticatedAt = null,
      refreshedAt = null;

  /// Whether currently authenticated
  bool get isAuthenticated => state == AuthState.authenticated;

  /// Whether an access token is available (even if expired)
  bool get hasTokens => accessToken != null;

  /// Whether authentication is in progress
  bool get isLoading =>
      state == AuthState.authenticating || state == AuthState.refreshing;

  /// Whether access token needs refresh
  bool get needsRefresh {
    if (!hasTokens || claims == null) return false;
    return claims!.willExpireWithin(const Duration(minutes: 2));
  }

  /// Copy with new values
  AuthStatus copyWith({
    AuthState? state,
    String? accessToken,
    TokenClaims? claims,
    User? userInfo,
    String? error,
    Map<String, dynamic>? metadata,
    DateTime? authenticatedAt,
    DateTime? refreshedAt,
  }) {
    return AuthStatus(
      state: state ?? this.state,
      accessToken: accessToken ?? this.accessToken,
      claims: claims ?? this.claims,
      userInfo: userInfo ?? this.userInfo,
      error: error ?? this.error,
      metadata: metadata ?? this.metadata,
      authenticatedAt: authenticatedAt ?? this.authenticatedAt,
      refreshedAt: refreshedAt ?? this.refreshedAt,
    );
  }

  @override
  List<Object?> get props => [
    state,
    accessToken,
    claims,
    userInfo,
    error,
    metadata,
    authenticatedAt,
    refreshedAt,
  ];

  @override
  String toString() {
    return 'AuthStatus(state: $state, hasTokens: $hasTokens, error: $error)';
  }
}

File: web-sdks/javascript/src/auth/auth_models.js
/**
 * Authentication models for GroupVAN SDK
 *
 * Models the JWT-based authentication system used by GroupVAN API.
 */

/**
 * Login request model
 */
export class LoginRequest {
  /**
   * @param {Object} options
   * @param {string} options.email - Email for authentication
   * @param {string} options.password - Password for authentication
   */
  constructor({ email, password }) {
    this.email = email;
    this.password = password;
  }

  toJson() {
    return {
      email: this.email,
      password: this.password,
    };
  }

  static fromJson(json) {
    return new LoginRequest({
      email: json.email,
      password: json.password,
    });
  }
}

/**
 * Token response from authentication endpoints
 */
export class TokenResponse {
  /**
   * @param {Object} options
   * @param {string} options.accessToken - JWT access token for API requests
   * @param {string} options.refreshToken - JWT refresh token for obtaining new access tokens
   * @param {number} options.expiresIn - Token expiration time in seconds
   * @param {string} [options.tokenType='Bearer'] - Token type
   */
  constructor({ accessToken, refreshToken, expiresIn, tokenType = 'Bearer' }) {
    this.accessToken = accessToken;
    this.refreshToken = refreshToken;
    this.expiresIn = expiresIn;
    this.tokenType = tokenType;
  }

  toJson() {
    return {
      access_token: this.accessToken,
      refresh_token: this.refreshToken,
      expires_in: this.expiresIn,
      token_type: this.tokenType,
    };
  }

  static fromJson(json) {
    return new TokenResponse({
      accessToken: json.access_token,
      refreshToken: json.refresh_token,
      expiresIn: json.expires_in,
      tokenType: json.token_type || 'Bearer',
    });
  }
}

/**
 * Refresh token request
 */
export class RefreshTokenRequest {
  /**
   * @param {Object} options
   * @param {string} options.refreshToken - Refresh token to use for generating new access token
   */
  constructor({ refreshToken }) {
    this.refreshToken = refreshToken;
  }

  toJson() {
    return {
      refresh_token: this.refreshToken,
    };
  }

  static fromJson(json) {
    return new RefreshTokenRequest({
      refreshToken: json.refresh_token,
    });
  }
}

/**
 * Logout request
 */
export class LogoutRequest {
  /**
   * @param {Object} options
   * @param {string} options.refreshToken - Refresh token to blacklist
   */
  constructor({ refreshToken }) {
    this.refreshToken = refreshToken;
  }

  toJson() {
    return {
      refresh_token: this.refreshToken,
    };
  }

  static fromJson(json) {
    return new LogoutRequest({
      refreshToken: json.refresh_token,
    });
  }
}

/**
 * JWT token claims decoded from access token
 */
export class TokenClaims {
  /**
   * @param {Object} options
   * @param {string} options.userId - User ID from the token
   * @param {string} [options.type='access'] - Token type (access/refresh)
   * @param {number} options.issuedAt - Token issued at timestamp (seconds)
   * @param {number} options.expiration - Token expiration timestamp (seconds)
   * @param {string} options.jti - Token JTI (unique identifier)
   * @param {string} [options.member] - Member ID if present
   */
  constructor({ userId, type = 'access', issuedAt, expiration, jti, member = null }) {
    this.userId = userId;
    this.type = type;
    this.issuedAt = issuedAt;
    this.expiration = expiration;
    this.jti = jti;
    this.member = member;
  }

  /**
   * Whether the token is expired
   * @returns {boolean}
   */
  get isExpired() {
    return Date.now() > this.expiration * 1000;
  }

  /**
   * Whether the token will expire within the given duration
   * @param {number} durationMs - Duration in milliseconds
   * @returns {boolean}
   */
  willExpireWithin(durationMs) {
    const expirationTime = new Date(this.expiration * 1000);
    const threshold = new Date(Date.now() + durationMs);
    return expirationTime < threshold;
  }

  /**
   * Time until token expires in milliseconds
   * @returns {number}
   */
  get timeUntilExpirationMs() {
    const expirationTime = this.expiration * 1000;
    const now = Date.now();
    if (expirationTime < now) {
      return 0;
    }
    return expirationTime - now;
  }

  toJson() {
    return {
      sub: this.userId,
      type: this.type,
      iat: this.issuedAt,
      exp: this.expiration,
      jti: this.jti,
      member: this.member,
    };
  }

  static fromJson(json) {
    return new TokenClaims({
      userId: json.sub,
      type: json.type || 'access',
      issuedAt: json.iat,
      expiration: json.exp,
      jti: json.jti,
      member: json.member,
    });
  }
}

/**
 * Authentication state
 * @enum {string}
 */
export const AuthState = {
  UNAUTHENTICATED: 'unauthenticated',
  AUTHENTICATING: 'authenticating',
  AUTHENTICATED: 'authenticated',
  REFRESHING: 'refreshing',
  EXPIRED: 'expired',
  FAILED: 'failed',
};

/**
 * Current authentication status
 */
export class AuthStatus {
  /**
   * @param {Object} options
   * @param {string} options.state - Current authentication state
   * @param {string} [options.accessToken] - Access token if authenticated
   * @param {string} [options.refreshToken] - Refresh token if authenticated
   * @param {TokenClaims} [options.claims] - Decoded token claims if available
   * @param {Object} [options.userInfo] - User info object
   * @param {string} [options.error] - Last authentication error if any
   * @param {Object} [options.metadata] - Additional metadata
   * @param {Date} [options.authenticatedAt] - Timestamp of last successful authentication
   * @param {Date} [options.refreshedAt] - Timestamp of last token refresh
   */
  constructor({
    state,
    accessToken = null,
    refreshToken = null,
    claims = null,
    userInfo = null,
    error = null,
    metadata = null,
    authenticatedAt = null,
    refreshedAt = null,
  }) {
    this.state = state;
    this.accessToken = accessToken;
    this.refreshToken = refreshToken;
    this.claims = claims;
    this.userInfo = userInfo;
    this.error = error;
    this.metadata = metadata;
    this.authenticatedAt = authenticatedAt;
    this.refreshedAt = refreshedAt;
  }

  /**
   * Create unauthenticated status
   * @returns {AuthStatus}
   */
  static unauthenticated() {
    return new AuthStatus({ state: AuthState.UNAUTHENTICATED });
  }

  /**
   * Create authenticating status
   * @returns {AuthStatus}
   */
  static authenticating() {
    return new AuthStatus({ state: AuthState.AUTHENTICATING });
  }

  /**
   * Create authenticated status
   * @param {Object} options
   * @param {string} options.accessToken
   * @param {string} options.refreshToken
   * @param {TokenClaims} options.claims
   * @param {Object} [options.userInfo]
   * @returns {AuthStatus}
   */
  static authenticated({ accessToken, refreshToken, claims, userInfo = null }) {
    return new AuthStatus({
      state: AuthState.AUTHENTICATED,
      accessToken,
      refreshToken,
      claims,
      userInfo,
      authenticatedAt: new Date(),
    });
  }

  /**
   * Create refreshing status
   * @param {Object} options
   * @param {string} options.accessToken
   * @param {string} options.refreshToken
   * @param {TokenClaims} options.claims
   * @param {Date} options.authenticatedAt
   * @param {Object} [options.userInfo]
   * @returns {AuthStatus}
   */
  static refreshing({ accessToken, refreshToken, claims, authenticatedAt, userInfo = null }) {
    return new AuthStatus({
      state: AuthState.REFRESHING,
      accessToken,
      refreshToken,
      claims,
      userInfo,
      authenticatedAt,
    });
  }

  /**
   * Create expired status
   * @param {Object} options
   * @param {string} [options.error]
   * @param {string} options.accessToken
   * @param {string} options.refreshToken
   * @param {Date} options.authenticatedAt
   * @param {Date} [options.refreshedAt]
   * @param {Object} [options.userInfo]
   * @returns {AuthStatus}
   */
  static expired({ error = null, accessToken, refreshToken, authenticatedAt, refreshedAt = null, userInfo = null }) {
    return new AuthStatus({
      state: AuthState.EXPIRED,
      accessToken,
      refreshToken,
      userInfo,
      error,
      authenticatedAt,
      refreshedAt,
    });
  }

  /**
   * Create failed status
   * @param {Object} options
   * @param {string} options.error
   * @param {Object} [options.metadata]
   * @returns {AuthStatus}
   */
  static failed({ error, metadata = null }) {
    return new AuthStatus({
      state: AuthState.FAILED,
      error,
      metadata,
    });
  }

  /**
   * Whether currently authenticated
   * @returns {boolean}
   */
  get isAuthenticated() {
    return this.state === AuthState.AUTHENTICATED;
  }

  /**
   * Whether tokens are available (even if expired)
   * @returns {boolean}
   */
  get hasTokens() {
    return this.accessToken !== null && this.refreshToken !== null;
  }

  /**
   * Whether authentication is in progress
   * @returns {boolean}
   */
  get isLoading() {
    return this.state === AuthState.AUTHENTICATING || this.state === AuthState.REFRESHING;
  }

  /**
   * Whether access token needs refresh
   * @returns {boolean}
   */
  get needsRefresh() {
    if (!this.hasTokens || !this.claims) return false;
    return this.claims.willExpireWithin(2 * 60 * 1000); // 2 minutes
  }

  /**
   * Copy with new values
   * @param {Object} overrides
   * @returns {AuthStatus}
   */
  copyWith(overrides = {}) {
    return new AuthStatus({
      state: overrides.state ?? this.state,
      accessToken: overrides.accessToken ?? this.accessToken,
      refreshToken: overrides.refreshToken ?? this.refreshToken,
      claims: overrides.claims ?? this.claims,
      userInfo: overrides.userInfo ?? this.userInfo,
      error: overrides.error ?? this.error,
      metadata: overrides.metadata ?? this.metadata,
      authenticatedAt: overrides.authenticatedAt ?? this.authenticatedAt,
      refreshedAt: overrides.refreshedAt ?? this.refreshedAt,
    });
  }

  toString() {
    return `AuthStatus(state: ${this.state}, hasTokens: ${this.hasTokens}, error: ${this.error})`;
  }
}
